[HBO-9]

NiakhaichykVitaliy opened the pull request.

NiakhaichykVitaliy committed:
[HBO-9] Add in dependencies retrofit and gson libs
Create PokemonApi interface
Use retrofit model in PokemonViewModel

NiakhaichykVitaliy committed:
[HBO-9] Delete log

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/viewmodel/PokemonViewModel.kt:
```diff
@@ -1,98 +1,44 @@
+import retrofit2.converter.gson.GsonConverterFactory
 
-class PokemonViewModel : ViewModel() {
+private const val BASE_URL = "https://mocki.io/"
```
У тебя в следующей задаче добавление нового покемона. Потенциально это тоже может быть через сервак. Поэтому это должен быть отдельный класс, который ты просто дергаешь, а он внутри создаёт ретрофит, отправляет запрос и т.д. и потом этот класс можно использовать в разных вьюмоделях или других классах

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/api/PokemonApi.kt:
```diff
@@ -0,0 +1,10 @@
+
+interface PokemonApi {
+    @GET("v1/76ac942a-a183-4522-87a9-16c2ef295a85")
+    fun getPokemon(): Call<List<Pokemon>>
```
Формат ответа сервера не всегда совпадает с доменными модельками, которые используются в коде в бизнес логике. Поэтому по хорошему лучше делать отдельные модельки для запросов и отдельные для домена. Тогда при изменении каких то вещей в модельки запроса не обязательно придется менять кучу кода по всему проекту, надо будет только поменять маппинг модельки запроса в модельку домена. 

Моделька запроса у тебя уже есть. Добавь для запроса и класс-маппер из одной модельки в другую (да, я понимаю, что в данном конкретном случае это будет супер-тупой маппинг одних и тех же данных с одинаковыми названиями. Но лучше, чтобы ты сразу привыкал так делать. В будущем такое разделение может очень помочь при изменениях на сервере)

NiakhaichykVitaliy committed:
[HBO-9] Delete import log
Create singleton RetrofitClient

NiakhaichykVitaliy committed:
[HBO-9] Create PokemonResponse class
Create Mapper for transform domain pokemonModel from server to pokemonModel app

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/api/PokemonMapper.kt:
```diff
@@ -0,0 +1,16 @@
+            Pokemon(
+                id = response.id,
+                name = response.name,
+                image = response.image
```
после последнего параметра тоже запятая

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/viewmodel/PokemonViewModel.kt:
```diff
@@ -3,96 +3,34 @@ package com.example.pokemon.viewmodel
-            )
-        )
-        _pokemonList.value = pokemonList
+    val pokemonApi: PokemonApi = RetrofitClient.retrofit.create(PokemonApi::class.java)
```
у тебя всё ещё ретрофит создается внутри вьюмодели. если надо будет его использовать где то ещё (в другой вьюмодели), то придется опять его вот так создавать. вынеси и создание ретрофита и работу маппера и обработку запрос в отдельный класс. чтобы во вьюмодели только дернуть одну функцию и получить результат

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/api/PokemonMapper.kt:
```diff
@@ -0,0 +1,16 @@
+import com.example.pokemon.model.Pokemon
+import com.example.pokemon.model.PokemonResponse
+
+object PokemonMapper {
```
почему маппер синглтон?

NiakhaichykVitaliy committed:
[HBO-9] made mapper not a singleton
moved fetchPokemonList to PokemonRepository class

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/viewmodel/PokemonViewModel.kt:
```diff
@@ -3,96 +3,28 @@ package com.example.pokemon.viewmodel
-            )
-        )
-        _pokemonList.value = pokemonList
+    private val pokemonApi: PokemonApi = RetrofitClient.retrofit.create(PokemonApi::class.java)
```
еще раз, ретрофитовский клиент не должен создаваться внутри какой то вьюмодели. вьюмодель вообще не должна знать, что ты используешь ретрофит - это слишком большая связанность. если ты вдруг решишь потом перейти с ретрофита на какую то другую библиотеку для получения данных с сервера, то тебе придётся менять код во всех вьюмоделях, которые используют ретрофит - так делать плохо.

создай отдельный класс, например PokemonNetwork или PokemonRemoteSource, и внутри него создавай ретрофитовский клиент. а во вьюмодели просто вызывай одну функцию

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/api/PokemonRepository.kt:
```diff
@@ -0,0 +1,33 @@
+            }
+
+            override fun onFailure(call: Call<List<PokemonResponse>>, t: Throwable) {
+                t.printStackTrace()
```
почему тут не вызывается колбек?

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/api/PokemonRepository.kt:
```diff
@@ -0,0 +1,33 @@
+
+class PokemonRepository(private val pokemonApi: PokemonApi) {
+
+    private val pokemonMapper: PokemonMapper by lazy { PokemonMapper() }
```
привыкай стараться прокидывать параметры через конструктор класса, а не создавать из внутри глобальными переменными - это потом поможет при тестировании и интеграции DI

NiakhaichykVitaliy committed:
[HBO-10] removed retrofit in PokemonRepository
removed val mapper inside method

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/ui/PokemonAdapter.java:
```diff
@@ -1,5 +1,6 @@
 package com.example.pokemon.ui;
 
+import android.util.Log;
```
это зачем?

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/api/PokemonRepository.kt:
```diff
@@ -0,0 +1,36 @@
+
+class PokemonRepository() {
+
+    private val pokemonApi: PokemonApi = RetrofitClient.retrofit.create(PokemonApi::class.java)
```
нет, репозиторий - это не то место, где должен создаваться ретрофит. про репозиторий есть отдельная задача. репозиторий - это как бы объединение разных объектов, которые могут предоставлять данные (БД, сервер, т.д.). но эти объекты должны быть абстрактными, а не привязанными к конкретной реализации

чтобы тебе было понятнее, давай сделаем разделение на интерфейсы и их реализации. сделай интерфейс, например `PokemonRemoteSource`, и для него реализация `PokemonRetrofitRemoteSource`. в реализации создай клиент ретрофита, вызови апи и маппер. а во вьюмодели прокинь интерфейс, который будет равен реализации

NiakhaichykVitaliy committed:
[HBO-10] delete import log

NiakhaichykVitaliy committed:
[HBO-9] Create interface PokemonRemoteSource, create PokemonRetrofitRemoteSource when realize interface PokemonRemoteSource
Refactor PokemonRepository (used PokemonRemoteSource)
Update PokemonViewModel

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/viewmodel/PokemonViewModel.kt:
```diff
@@ -3,96 +3,27 @@ package com.example.pokemon.viewmodel
+            }
+            result.onFailure { throwable ->
+                throwable.printStackTrace()
+                _pokemonList.value = emptyList()
```
а это зачем?

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/api/remote/PokemonRetrofitRemoteSource.kt:
```diff
@@ -0,0 +1,38 @@
+import retrofit2.Response
+
+class PokemonRetrofitRemoteSource : PokemonRemoteSource {
+    private val pokemonApi: PokemonApi = RetrofitClient.retrofit.create(PokemonApi::class.java)
```
давай сделаем так:

```
object RetrofitClient {
    val pokemonApi: PokemonApi by lazy { RetrofitClient.retrofit.create(PokemonApi::class.java) }
    ...
}

class PokemonRetrofitRemoteSource(
    private val pokemonApi: PokemonApi = RetrofitClient.pokemonApi,
) : PokemonRemoteSource { ... }
```
чтобы зависимости не создавались внутри класса, а прокидывались через конструктор. это поможет потом в интеграции DI и написании тестов

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/api/PokemonRepository.kt:
```diff
@@ -0,0 +1,10 @@
+
+import com.example.pokemon.model.Pokemon
+
+class PokemonRepository(private val pokemonRemoteSource: PokemonRemoteSource) {
```
давай тут сразу проинициализируем дефолтным значением
```
class PokemonRepository(
    private val pokemonRemoteSource: PokemonRemoteSource = PokemonRetrofitRemoteSource(),
) { ... }
```

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/api/remote/PokemonRetrofitRemoteSource.kt:
```diff
@@ -0,0 +1,38 @@
+                response: Response<List<PokemonResponse>>
+            ) {
+                if (response.isSuccessful && response.body() != null) {
+                    val pokemonMapper = PokemonMapper()
```
маппер тоже в конструктор

DerevyagoKonstantin commented on pokemon/src/main/java/com/example/pokemon/viewmodel/PokemonViewModel.kt:
```diff
@@ -3,96 +3,27 @@ package com.example.pokemon.viewmodel
     private val _pokemonList = MutableLiveData<List<Pokemon>>()
     val pokemonList: LiveData<List<Pokemon>> = _pokemonList
+    private val pokemonRemoteSource: PokemonRemoteSource = PokemonRetrofitRemoteSource()
+    private val pokemonRepository = PokemonRepository(pokemonRemoteSource)
```
репозиторий в конструктор
